File: GameServer/Common/Packet.cpp
#include "Packet.h"

namespace Packet
{
	bool CJsonSerializer::Serialize( IJsonSerializable* pObj, std::string& output )
	{
		if (pObj == NULL)
			return false;

		Json::Value serializeRoot;
		pObj->Serialize(serializeRoot);

		Json::StyledWriter writer;
		output = writer.write( serializeRoot );

		return true;
	}

	bool CJsonSerializer::Deserialize( IJsonSerializable* pObj, std::string& input )
	{
		if (pObj == NULL)
			return false;

		Json::Value deserializeRoot;
		Json::Reader reader;

		if ( !reader.parse(input, deserializeRoot) )
			return false;

		pObj->Deserialize(deserializeRoot);

		return true;
	}

	void LoginReq::Serialize( Json::Value& root )
	{
		root["_id"] = _id;
		root["_token"] = _token;
	}

	void LoginReq::Deserialize( Json::Value& root )
	{
		_id = root.get("_id", "").asString();
		_token = root.get("_token", "").asString();
	}

	void LoginRes::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void LoginRes::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void FastMatchReq::Serialize( Json::Value& root )
	{
		root["_type"] = _type;
		root["_id"] = _id;
		root["_token"] = _token;
	}

	void FastMatchReq::Deserialize( Json::Value& root )
	{
		_type = root.get("_type", 0).asInt();
		_id = root.get("_id", "").asString();
		_token = root.get("_token", "").asString();
	}

	void FastMatchRes::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void FastMatchRes::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void MatchCancelReq::Serialize( Json::Value& root )
	{
		root["_id"] = _id;
		root["_token"] = _token;
	}

	void MatchCancelReq::Deserialize( Json::Value& root )
	{
		_id = root.get("_id", "").asString();
		_token = root.get("_token", "").asString();
	}

	void MatchCancelRes::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void MatchCancelRes::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void MatchSuccessNotify::Serialize( Json::Value& root )
	{
		root["_gameNumber"] = _gameNumber;
		root["_enemyId"] = _enemyId;
		root["_enemyWins"] = _enemyWins;
		root["_enemyLoses"] = _enemyLoses;
		root["_enemyType"] = _enemyType;
	}

	void MatchSuccessNotify::Deserialize( Json::Value& root )
	{
		_gameNumber = root.get("_gameNumber", 0).asInt();
		_enemyId = root.get("_enemyId", "").asString();
		_enemyWins = root.get("_enemyWins", 0).asInt();
		_enemyLoses = root.get("_enemyLoses", 0).asInt();
		_enemyType = root.get("_enemyType", 0).asInt();
	}

	void MatchSuccessAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void MatchSuccessAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void GameStartNotify::Serialize( Json::Value& root )
	{
		root["_playerNumber"] = _playerNumber;
		root["_positionX"] = _positionX;
		root["_enemyPositionX"] = _enemyPositionX;
	}

	void GameStartNotify::Deserialize( Json::Value& root )
	{
		_playerNumber = root.get("_playerNumber", 0).asInt();
		_positionX = root.get("_positionX", 0.0).asDouble();
		_enemyPositionX = root.get("_enemyPositionX", 0.0).asDouble();
	}

	void GameStartAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void GameStartAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void TurnStartNotify::Serialize( Json::Value& root )
	{
		root["_windX"] = _windX;
		root["_windY"] = _windY;
	}

	void TurnStartNotify::Deserialize( Json::Value& root )
	{
		_windX = root.get("_windX", 0.0).asDouble();
		_windY = root.get("_windY", 0.0).asDouble();
	}

	void TurnStartAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void TurnStartAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void EnemyTurnStartNotify::Serialize( Json::Value& root )
	{
		root["_windX"] = _windX;
		root["_windY"] = _windY;
	}

	void EnemyTurnStartNotify::Deserialize( Json::Value& root )
	{
		_windX = root.get("_windX", 0.0).asDouble();
		_windY = root.get("_windY", 0.0).asDouble();
	}

	void EnemyTurnStartAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void EnemyTurnStartAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void MoveNotify::Serialize( Json::Value& root )
	{
		root["_enemyPositionX"] = _enemyPositionX;
	}

	void MoveNotify::Deserialize( Json::Value& root )
	{
		_enemyPositionX = root.get("_enemyPositionX", 0.0).asDouble();
	}

	void MoveAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void MoveAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void EnemyMoveNotify::Serialize( Json::Value& root )
	{
		root["_enemyPositionX"] = _enemyPositionX;
	}

	void EnemyMoveNotify::Deserialize( Json::Value& root )
	{
		_enemyPositionX = root.get("_enemyPositionX", 0.0).asDouble();
	}

	void EnemyMoveAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void EnemyMoveAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void FireNotify::Serialize( Json::Value& root )
	{
		root["_fireType"] = _fireType;
		root["_enemyPositionX"] = _enemyPositionX;
		root["_unitVecX"] = _unitVecX;
		root["_unitVecY"] = _unitVecY;
		root["_magnitude"] = _magnitude;
	}

	void FireNotify::Deserialize( Json::Value& root )
	{
		_fireType = root.get("_fireType", 0).asInt();
		_enemyPositionX = root.get("_enemyPositionX", 0.0).asDouble();
		_unitVecX = root.get("_unitVecX", 0.0).asDouble();
		_unitVecY = root.get("_unitVecY", 0.0).asDouble();
		_magnitude = root.get("_magnitude", 0.0).asDouble();
	}

	void FireAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void FireAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void EnemyFireNotify::Serialize( Json::Value& root )
	{
		root["_fireType"] = _fireType;
		root["_enemyPositionX"] = _enemyPositionX;
		root["_unitVecX"] = _unitVecX;
		root["_unitVecY"] = _unitVecY;
		root["_magnitude"] = _magnitude;
	}

	void EnemyFireNotify::Deserialize( Json::Value& root )
	{
		_fireType = root.get("_fireType", 0).asInt();
		_enemyPositionX = root.get("_enemyPositionX", 0.0).asDouble();
		_unitVecX = root.get("_unitVecX", 0.0).asDouble();
		_unitVecY = root.get("_unitVecY", 0.0).asDouble();
		_magnitude = root.get("_magnitude", 0.0).asDouble();
	}

	void EnemyFireAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void EnemyFireAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void TurnEndNotify::Serialize( Json::Value& root )
	{
	}

	void TurnEndNotify::Deserialize( Json::Value& root )
	{
	}

	void TurnEndAck::Serialize( Json::Value& root )
	{
	}

	void TurnEndAck::Deserialize( Json::Value& root )
	{
	}

	void GameSetNotify::Serialize( Json::Value& root )
	{
		root["_winPlayerNum"] = _winPlayerNum;
	}

	void GameSetNotify::Deserialize( Json::Value& root )
	{
		_winPlayerNum = root.get("_winPlayerNum", 0).asInt();
	}

	void GameSetAck::Serialize( Json::Value& root )
	{
		root["_result"] = _result;
	}

	void GameSetAck::Deserialize( Json::Value& root )
	{
		_result = root.get("_result", 0).asInt();
	}

	void CloseReq::Serialize( Json::Value& root )
	{
		root["_id"] = _id;
		root["_token"] = _token;
	}

	void CloseReq::Deserialize( Json::Value& root )
	{
		_id = root.get("_id", "").asString();
		_token = root.get("_token", "").asString();
	}

}

File: Common/Packet.cs
namespace PacketInfo
{
	// C++ IOCP서버와 Unity C#이 통신하기 위한 패킷 정의 파일입니다.
	// 로그인 요청 패킷
	public class LoginReq
{
	public string				_id;
	public string				_token;
	}

	// 로그인 답변 패킷
	public class LoginRes
{
	public int					_result;
	}

	// 빠른 매치 요청 패킷
	public class FastMatchReq
{
	public int					_type;
	public string				_id;
	public string				_token;
	}

	// 매치 요청 답변 패킷
	public class FastMatchRes
{
	public int					_result;
	}

	// 매치 취소 패킷
	public class MatchCancelReq
{
	public string				_id;
	public string				_token;
	}

	// 매치 취소 답변 패킷
	public class MatchCancelRes
{
	public int					_result;
	}

	// 매칭 성사 알림 패킷
	public class MatchSuccessNotify
{
	public int					_gameNumber;
	public string				_enemyId;
	public int					_enemyWins;
	public int					_enemyLoses;
	public int					_enemyType;
	}

	// 매칭 성사 응답 패킷
	public class MatchSuccessAck
{
	public int					_result;
	}

	// 게임 시작 알림 패킷
	public class GameStartNotify
{
	public int					_playerNumber;
	public float				_positionX;
	public float				_enemyPositionX;
	}

	// 게임 시작 응답 패킷
	public class GameStartAck
{
	public int					_result;
	}

	// 내 턴 시작 패킷
	public class TurnStartNotify
{
	public float				_windX;
	public float				_windY;
	}

	// 내 턴 시작 응답 패킷
	public class TurnStartAck
{
	public int					_result;
	}

	// 상대 턴 시작 패킷
	public class EnemyTurnStartNotify
{
	public float				_windX;
	public float				_windY;
	}

	// 상대 턴 시작 응답 패킷
	public class EnemyTurnStartAck
{
	public int					_result;
	}

	// 움직임 전달 패킷
	public class MoveNotify
{
	public float				_enemyPositionX;
	public float				_enemyPositionY;
	}

	// 움직임 응답 패킷
	public class MoveAck
{
	public int					_result;
	}

	// 상대 움직임 알림 패킷
	public class EnemyMoveNotify
{
	public float				_enemyPositionX;
	public float				_enemyPositionY;
	}

	// 상대 움직임 응답 패킷
	public class EnemyMoveAck
{
	public int					_result;
	}

	// 발사 패킷
	public class FireNotify
{
	public int					_fireType;
	public int					_damage;
	public float				_enemyPositionX;
	public float				_enemyPositionY;
	public float				_unitVecX;
	public float				_unitVecY;
	public float				_magnitude;
	}

	// 발사 응답 패킷
	public class FireAck
{
	public int					_result;
	}

	// 상대 발사 패킷
	public class EnemyFireNotify
{
	public int					_fireType;
	public int					_damage;
	public float				_enemyPositionX;
	public float				_enemyPositionY;
	public float				_unitVecX;
	public float				_unitVecY;
	public float				_magnitude;
	}

	// 상대 발사 응답 패킷
	public class EnemyFireAck
{
	public int					_result;
	}

	// 턴 종료 알림 패킷
	public class TurnEndNotify
{
	}

	// 턴 종료 응답 패킷
	public class TurnEndAck
{
	}

	// 데미지가 발생할때 보내는 패킷
	public class DamageOccur
{
	public int					_player1Hp;
	public int					_player2Hp;
	}

	// 게임 종료 알림 패킷
	public class GameSetNotify
{
	public int					_isGameSet;
	public int					_player1Hp;
	public int					_player2Hp;
	public int					_winPlayerNum;
	}

	// 게임 종료 응답 패킷
	public class GameSetAck
{
	public int					_result;
	}

	// 접속 닫기 패킷
	public class CloseReq
{
	public string				_id;
	public string				_token;
	}

	public enum PacketId
	{
		ID_LoginReq			= 101,
		ID_LoginRes			= 102,
		ID_FastMatchReq			= 103,
		ID_FastMatchRes			= 104,
		ID_MatchCancelReq			= 105,
		ID_MatchCancelRes			= 106,
		ID_MatchSuccessNotify			= 107,
		ID_MatchSuccessAck			= 108,
		ID_GameStartNotify			= 109,
		ID_GameStartAck			= 110,
		ID_TurnStartNotify			= 111,
		ID_TurnStartAck			= 112,
		ID_EnemyTurnStartNotify			= 113,
		ID_EnemyTurnStartAck			= 114,
		ID_MoveNotify			= 115,
		ID_MoveAck			= 116,
		ID_EnemyMoveNotify			= 117,
		ID_EnemyMoveAck			= 118,
		ID_FireNotify			= 119,
		ID_FireAck			= 120,
		ID_EnemyFireNotify			= 121,
		ID_EnemyFireAck			= 122,
		ID_TurnEndNotify			= 123,
		ID_TurnEndAck			= 124,
		ID_DamageOccur			= 125,
		ID_GameSetNotify			= 126,
		ID_GameSetAck			= 127,
		ID_CloseReq			= 128,
	};
}